### PR TITLE
KAA-42: Add handling of client session expiration with optional redirection to login page.

### PR DESCRIPTION
**Pull request by ikulikov**

### CODE REVIEW

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/activity/AbstractDetailsActivity.java [hunk 1]
```diff
@@ -157,10 +160,19 @@ public void onSuccess(T result) {
...
                     }
         });
     }
 
+    @Override
+    public String mayStop() {
+        if (detailsView.hasChanged() && !canceled) {
+            return Utils.messages.detailsMayStopMessage();
+        }
+        else {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/activity/AbstractDetailsActivity.java]: Code style

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/activity/AbstractSchemaActivity.java [hunk 2]
```diff
@@ -117,10 +121,27 @@ public void onSuccess(T result) {
...
+                            Utils.handleException(caught, detailsView);
                         }
             });
         }
     }
+    
+    @Override
+    public String customizeErrorMessage(Throwable caught) {
+        String errorMessage = caught.getMessage();
+        if (errorMessage.contains(LEFT_SQUARE_BRACKET) && errorMessage.contains(RIGHT_SQUARE_BRACKET)) {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/activity/AbstractSchemaActivity.java]: I propose a bit more effective way:

```
    String errorMessage = caught.getMessage();

    int leftSquareBracketIndex = errorMessage.indexOf(LEFT_SQUARE_BRACKET);
    int rightSquareBracketIndex = -1;

    if (leftSquareBracketIndex != -1) {
        rightSquareBracketIndex = errorMessage.indexOf(RIGHT_SQUARE_BRACKET, leftSquareBracketIndex);
    }

    if (rightSquareBracketIndex != -1) {
        StringBuilder builder = new StringBuilder();
        builder.append("Incorrect json schema: Please check your schema at");
        String[] array = errorMessage.substring(leftSquareBracketIndex, rightSquareBracketIndex).split(SEMICOLON);
        if (array != null && array.length == 2) {
            builder.append(array[1]);
            errorMessage = builder.toString();
        }
    } else {
        return "Incorrect schema: Please validate your schema.";
    }
    return errorMessage;
```

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/util/Utils.java [hunk 3]
```diff
@@ -46,39 +50,64 @@
...
-            ServiceErrorCode errorCode = ((KaaAdminServiceException)throwable).getErrorCode();
+    public static void handleException(Throwable caught, HasErrorMessage hasErrorMessage) {
+        handleException(caught, hasErrorMessage, null);
+    }
+
+    public static void handleException(Throwable caught, HasErrorMessage hasErrorMessage, ErrorMessageCustomizer errorMessageCustomizer) {
+        boolean handled = false;
+        if (caught instanceof StatusCodeException) {
+            StatusCodeException sce = (StatusCodeException)caught;
+            if (sce.getStatusCode()==Response.SC_UNAUTHORIZED) {
```
Comment (shpigor) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/util/Utils.java]: Use code formatting, please.

**Comment by shpigor** [comment]
Reivewed

**Pull request merged**